##  0. Import library

In [2]:
import time
import numpy as np
from numpy.random import *
np.random.seed(0)

## 1. Prepare input data

In [3]:
npop = 10 # population size
sigma = 0.1 # noise standard deviation
alpha = 0.001 # learning rate

w = np.random.rand(3) # our initial guess is random
w = w/sum(w) # notmalization (sum of w =1)
w

array([ 0.29399155,  0.38311672,  0.32289173])

In [4]:
cov = np.array([(6,-5,4), (-5,17,-11),(4,-11,24)])  # covariance matrix of the portfolio
mu=np.array([8,12,15])                                               #  avarage return of each asset in the portfolio
num=len(mu)
beta=2

## 2. Create the functions

### 2.1 The function for evaluate actions

In [5]:
# The function for evaluate actions
def f(w,mu,cov,beta):
     preturn=np.dot(mu, w)
     pvar=0.5*beta*(np.dot(np.dot(w, cov), w))
     reward=preturn-pvar
   
     return reward

### 2.2 Create initial populations and reward matrix

In [6]:
# nitial populations and reward matrix
N = np.random.randn(npop, 3) # samples from a normal distribution N(0,1)
R = np.zeros(npop)
N,R

(array([[-2.2683282 ,  1.33354538, -0.84272405],
        [ 1.96992445,  1.26611853, -0.50587654],
        [ 2.54520078,  1.08081191,  0.48431215],
        [ 0.57914048, -0.18158257,  1.41020463],
        [-0.37447169,  0.27519832, -0.96075461],
        [ 0.37692697,  0.03343893,  0.68056724],
        [-1.56349669, -0.56669762, -0.24214951],
        [ 1.51439128, -0.3330574 ,  0.04736482],
        [ 1.46274045,  1.53502913,  0.56644004],
        [ 0.14926509, -1.078278  ,  1.39547227]]),
 array([ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.]))

### 2.3 Inject noise into parameters

In [7]:
w_try = w + sigma*N[0] # jitter w using gaussian of sigma 0.1
R[0] = f(w_try,mu,cov,beta) # evaluate the jittered version
w_try,R

(array([ 0.06715873,  0.51647126,  0.23861932]),
 array([ 7.31592217,  0.        ,  0.        ,  0.        ,  0.        ,
         0.        ,  0.        ,  0.        ,  0.        ,  0.        ]))

## 3. Train the model

In [9]:
# Train the model
t=time.time()
for i in range(100000):

  # initialize memory for a population of w's, and their rewards
  N = np.random.randn(npop, 3) # samples from a normal distribution N(0,1)
  R = np.zeros(npop)
  for j in range(npop):
    w_try = w + sigma*N[j] # inject noise using gaussian of sigma 0.1
    R[j] = f(w_try,mu,cov,beta) # evaluate the function after injecting noise to parameters

  # standardize the rewards to have a gaussian distribution
  A = (R - np.mean(R)) / np.std(R)
  # perform the parameter update. 
  w = w + alpha/(npop*sigma) * np.dot(N.T, A)
  w = w/sum(w)
    
t2 = time.time()
print(round(t2-t, 5), 'Seconds to predict')
    

Return=np.dot(w, mu)                                    # caluculate avarage return of the portfolio
Risk=np.sqrt(np.dot(np.dot(w, cov), w))    # caluculate standard deviation of the portfolio
    
w,Return,Risk       
     


16.98766 Seconds to predict


(array([ 0.30050823,  0.38890967,  0.3105821 ]),
 11.729713408036291,
 1.5325708017339077)

Original : 0.239, 0.433, 0.326, 12.02, 1.604

### This code is solely for educational purpose. The code cannot be used for investments in practive. TOSHI STATS SDN. BHD. and I do not accept any responsibility or liability for loss or damage occasioned to any person or property through using materials, instructions, methods, algorithm or ideas contained herein, or acting or refraining from acting as a result of such use. TOSHI STATS SDN. BHD. and I expressly disclaim all implied warranties, including merchantability or fitness for any particular purpose. There will be no duty on TOSHI STATS SDN. BHD. and me to correct any errors or defects in the codes and the software.                    
